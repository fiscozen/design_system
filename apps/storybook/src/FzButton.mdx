import { Meta } from '@storybook/addon-docs/blocks';

<Meta title="Documentation/Button/FzButton" />

# FzButton

A comprehensive button component supporting five visual variants and two environment-based sizes, with full icon integration and accessibility support.

## Features

- **5 Variants**: Primary, Secondary, Invisible, Danger, Success
- **2 Environments**: Backoffice (compact), Frontoffice (larger - default)
- **Icon Support**: Before, after, or both positions
- **4 Interactive States**: Default, Hover, Focus, Disabled
- **Full Accessibility**: Keyboard navigation, ARIA support
- **Customizable**: Slots and class overrides

## Installation

```bash
pnpm add @fiscozen/button
```

## Basic Usage

```vue
<script setup lang="ts">
import { FzButton } from '@fiscozen/button'
</script>

<template>
  <FzButton label="Click me" />
</template>
```

## Props

<table>
  <thead>
    <tr>
      <th>Prop</th>
      <th>Type</th>
      <th>Default</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>label</code></td>
      <td><code>string</code></td>
      <td><code>undefined</code></td>
      <td>Text label displayed on the button</td>
    </tr>
    <tr>
      <td><code>variant</code></td>
      <td><code>'primary' | 'secondary' | 'invisible' | 'danger' | 'success'</code></td>
      <td><code>'primary'</code></td>
      <td>Visual style variant</td>
    </tr>
    <tr>
      <td><code>environment</code></td>
      <td><code>'backoffice' | 'frontoffice'</code></td>
      <td><code>'frontoffice'</code></td>
      <td>Button environment determining size and spacing</td>
    </tr>
    <tr>
      <td><code>disabled</code></td>
      <td><code>boolean</code></td>
      <td><code>false</code></td>
      <td>Whether the button is disabled</td>
    </tr>
    <tr>
      <td><code>iconName</code></td>
      <td><code>string</code></td>
      <td><code>undefined</code></td>
      <td>FontAwesome icon name (e.g., 'bell', 'save')</td>
    </tr>
    <tr>
      <td><code>iconVariant</code></td>
      <td><code>IconVariant</code></td>
      <td><code>undefined</code></td>
      <td>FontAwesome icon variant (fas, far, fal, etc.)</td>
    </tr>
    <tr>
      <td><code>iconPosition</code></td>
      <td><code>'before' | 'after'</code></td>
      <td><code>'before'</code></td>
      <td>Position of the icon relative to the label</td>
    </tr>
    <tr>
      <td><code>containerClass</code></td>
      <td><code>string</code></td>
      <td><code>undefined</code></td>
      <td>Additional CSS classes for the label container</td>
    </tr>
    <tr>
      <td><code>overrideContainerClass</code></td>
      <td><code>boolean</code></td>
      <td><code>false</code></td>
      <td>If true, replaces default container classes instead of merging</td>
    </tr>
  </tbody>
</table>

**Deprecated Props:**
- `size` (`'xs' | 'sm' | 'md' | 'lg'`): Use `environment` instead
- `tooltip` (`string`): Use `FzTooltip` component wrapper instead

## Slots

<table>
  <thead>
    <tr>
      <th>Slot</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>default</code></td>
      <td>Custom content for the button (overrides <code>label</code> prop)</td>
    </tr>
    <tr>
      <td><code>before</code></td>
      <td>Custom content before the label (replaces icon when provided)</td>
    </tr>
    <tr>
      <td><code>after</code></td>
      <td>Custom content after the label (replaces icon when provided)</td>
    </tr>
  </tbody>
</table>

## Variants

### Primary (Default)
Primary actions and main CTAs. Blue background with white text.

```vue
<FzButton variant="primary" label="Primary Action" />
```

### Secondary
Secondary actions. White background with grey text and border.

```vue
<FzButton variant="secondary" label="Secondary Action" />
```

### Invisible
Subtle actions. Transparent background with grey text.

```vue
<FzButton variant="invisible" label="Subtle Action" />
```

### Danger
Destructive actions. Red semantic color background.

```vue
<FzButton variant="danger" label="Delete" />
```

### Success
Confirmation actions. Green semantic color background.

```vue
<FzButton variant="success" label="Confirm" />
```

## With Icons

Buttons support icons before, after, or on both sides of the label. Icons are always 16px (md size) for consistency.

```vue
<template>
  <!-- Icon before label (default) -->
  <FzButton 
    label="Save" 
    iconName="save" 
    iconPosition="before" 
  />
  
  <!-- Icon after label -->
  <FzButton 
    label="Next" 
    iconName="arrow-right" 
    iconPosition="after" 
  />
  
  <!-- Both icons -->
  <FzButton label="Notification">
    <template #before>
      <FzIcon name="bell" />
    </template>
    <template #after>
      <FzIcon name="bell" />
    </template>
  </FzButton>
</template>
```

## Environments

### Backoffice
Compact size for internal applications.

```vue
<FzButton environment="backoffice" label="Backoffice Button" />
```

### Frontoffice (Default)
Larger size for customer-facing interfaces.

```vue
<FzButton environment="frontoffice" label="Frontoffice Button" />
```

## States

### Disabled
All variants support disabled state with lighter colors.

```vue
<FzButton label="Disabled Button" :disabled="true" />
```

## Special Cases

### Long Text with Truncation
Text automatically truncates with ellipsis when it exceeds button width.

```vue
<FzButton label="This is a very long button text that will be truncated with ellipsis when it exceeds the maximum width" />
```

## Interactive States

All button variants support four interactive states:

- **Default**: Base background color for each variant
- **Hover**: Darker background color on mouse over
- **Focus**: Visible colored border for keyboard navigation (no flickering)
- **Disabled**: Lighter background with reduced opacity

## Icon Behavior

- Icons only render when both label and iconName are provided
- Custom icons can be provided via `#before` and `#after` slots

## Accessibility

- Native `<button>` element with `type="button"`
- Full keyboard navigation (Tab, Enter, Space)
- Disabled state prevents interaction
- Focus states clearly visible with colored border
- Compatible with screen readers

## Deprecation Notices

### Size Prop
The `size` prop is deprecated. Use `environment` instead:

```vue
<!-- ❌ Deprecated -->
<FzButton size="lg" label="Button" />

<!-- ✅ Recommended -->
<FzButton environment="frontoffice" label="Button" />
```

**Size to Environment Mapping:**
- `xs`, `sm`, `md` → `backoffice`
- `lg` → `frontoffice`

### Tooltip Prop
The `tooltip` prop is deprecated. Use the `FzTooltip` component wrapper instead:

```vue
<!-- ❌ Deprecated -->
<FzButton tooltip="Click me" label="Button" />

<!-- ✅ Recommended -->
<FzTooltip content="Click me">
  <FzButton label="Button" />
</FzTooltip>
```

## Usage Examples

### Form Actions
```vue
<div class="flex gap-12">
  <FzButton variant="primary" label="Submit" @click="handleSubmit" />
  <FzButton variant="secondary" label="Cancel" @click="handleCancel" />
</div>
```

### Destructive Action
```vue
<FzButton 
  variant="danger" 
  label="Delete Account" 
  iconName="trash"
  :disabled="!confirmed"
  @click="handleDelete"
/>
```

### Navigation
```vue
<div class="flex gap-8">
  <FzButton 
    variant="secondary" 
    label="Back" 
    iconName="arrow-left" 
    environment="backoffice"
  />
  <FzButton 
    variant="primary" 
    label="Continue" 
    iconName="arrow-right" 
    iconPosition="after"
    environment="backoffice"
  />
</div>
```

### Custom Content with Slots
```vue
<FzButton>
  <span>Custom <strong>HTML</strong> Content</span>
</FzButton>

<FzButton label="Custom Icons">
  <template #before>
    <CustomIcon />
  </template>
  <template #after>
    <CustomIcon />
  </template>
</FzButton>
```

## Related Components

- **FzIconButton**: Icon-only button for compact interfaces
- **FzButtonGroup**: Container for grouping related buttons
- **FzTooltip**: Wrap buttons to add tooltips

