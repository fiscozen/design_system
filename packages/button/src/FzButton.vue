<template>
  <button type="button" :disabled="disabled" :class="[staticClasses, classes]">
    <div v-if="iconAndLabel" :class="[staticIconClasses, beforeClasses]">
      <slot name="before">
        <FzIcon
          v-if="iconName && iconPosition === 'before'"
          :name="iconName"
          :size="mappedIconSize"
          :variant="iconVariant"
        />
      </slot>
    </div>
    <div :class="containerClass">
      <slot>
        {{ label }}
      </slot>
    </div>
    <div v-if="slots.after || iconAndLabel" :class="[staticIconClasses, afterClasses]">
      <slot name="after">
        <FzIcon
          v-if="iconName && iconPosition === 'after'"
          :name="iconName"
          :size="mappedIconSize"
        />
      </slot>
    </div>
    <span class="hidden h-0 w-0">{{ tooltip }}</span>
  </button>
</template>

<script lang="ts" setup>
import { computed, useSlots } from 'vue'
import { type IconVariant, FzIcon } from '@fiscozen/icons'
import type { ButtonSize, ButtonVariant } from './types'
import { iconSizeMap } from './utils'

const props = withDefaults(
  defineProps<{
    /**
     * The label of the button
     */
    label?: string
    /**
     * The tooltip of the button will be shown on hover
     */
    tooltip?: string
    /**
     * primary or secondary button
     */
    variant?: ButtonVariant
    /**
     * size of the button
     */
    size?: ButtonSize
    /**
     * whether action is enabled or not
     */
    disabled?: boolean
    /**
     * Icon to be displayed. Use fontawesome search here https://fontawesome.com/v6/icons
     * Mind that not all of the icons and variants might be available.
     */
    iconName?: string
    /**
     * Fontawesome icon variant: solid, regular, light, thin. Sharp subvariants are available as well
     */
    iconVariant?: IconVariant
    /**
     * Positioning of the icon
     */
    iconPosition?: 'before' | 'after',
    /**
     * Define the css class for the default slot container
     */
    containerClass?: string
    /**
     * Whether to override internal container classes
     */
    overrideContainerClass?: boolean
  }>(),
  {
    variant: 'primary',
    size: 'md',
    disabled: false,
    iconPosition: 'before',
  }
)
console.log('FzButton', props)
const slots = useSlots()

const customVariantClasses = computed(() => {
  switch (props.variant) {
    case 'primary':
      return {
        'bg-blue-500': true,
        'hover:bg-blue-600': true,
        'disabled:bg-blue-200': true,
        'text-core-white': true,
        'focus:bg-blue-500': !props.disabled
      }
      break
    case 'secondary':
      return {
        'text-grey-500': true,
        'bg-core-white': true,
        '!border-grey-200': true,
        'hover:bg-grey-100': !props.disabled,
        'focus:!border-blue-600': !props.disabled,
        'disabled:text-grey-100': true,
      }
      break
    case 'invisible':
      return {
        'text-grey-500': true,
        'bg-transparent': true,
        'border-transparent': true,
        'hover:bg-grey-100': !props.disabled,
        'focus:!border-blue-600': !props.disabled,
        'disabled:text-grey-100': true
      }
      break
    case 'danger':
      return {
        'text-core-white': true,
        'temporary-bg-red': true,
        'temporary-border-red': !props.disabled,
        'bg-semantic-error': true
      }
      break
    case 'success':
      return {
        'text-core-white': true,
        'temporary-bg-green': true,
        'temporary-border-green': !props.disabled,
        'bg-semantic-success': true
      }
      break
    default:
      return {}
      break
  }
})

const iconAndLabel = computed(() => Boolean((props.label || slots.default) && props.iconName))
const mappedIconSize = computed(() => iconSizeMap[props.size])

const staticClasses = [
  'relative',
  'rounded',
  'flex',
  'flex-shrink',
  'items-center',
  'justify-center',
  'font-medium',
  'border-1',
  'border-transparent'
]

const staticIconClasses = ['flex', 'items-center', 'justify-items-center']

const classes = computed(() => ({
  'h-24 text-xs': props.size === 'xs',
  'h-28 text-sm': props.size === 'sm',
  'h-32': props.size === 'md',
  'h-40 text-lg': props.size === 'lg',
  'focus:border-blue-600': !props.disabled,
  'focus:border-solid': !props.disabled,
  'focus:border-1': !props.disabled,
  'px-12': props.size === 'xs' && !iconAndLabel.value,
  'px-14': props.size === 'sm' && !iconAndLabel.value,
  'px-16': props.size === 'md' && !iconAndLabel.value,
  'px-20': props.size === 'lg' && !iconAndLabel.value,
  ...customVariantClasses.value
}))

const beforeClasses = computed(() => ({
  'mr-8': props.size === 'lg',
  'mr-6': props.size === 'md',
  'mr-4': props.size === 'sm' || props.size === 'xs',
  'pl-12': props.size === 'lg' && iconAndLabel.value,
  'pl-10': (props.size === 'md' || props.size === 'sm') && iconAndLabel.value,
  'pl-8': props.size === 'xs' && iconAndLabel.value
}))

const afterClasses = computed(() => ({
  'ml-8': props.size === 'lg',
  'ml-6': props.size === 'md',
  'ml-4': props.size === 'sm' || props.size === 'xs',
  'pr-12': props.size === 'lg' && iconAndLabel.value,
  'pr-10': (props.size === 'md' || props.size === 'sm') && iconAndLabel.value,
  'pr-8': props.size === 'xs' && iconAndLabel.value
}))

const containerClass = computed(() => {
  if (props.overrideContainerClass) {
    return props.containerClass
  }

  return ["truncate", props.containerClass]
});
</script>

<style>
.temporary-bg-red:hover {
  background-color: #aa2f2f;
}
.temporary-bg-red:disabled {
  background-color: #f8baba;
}
.temporary-border-red:focus {
  border-color: #aa2f2f;
}
.temporary-bg-green:hover {
  background-color: #0b7763;
}
.temporary-bg-green:disabled {
  background-color: #b5d8ce;
}
.temporary-border-green:focus {
  border-color: #0b7763;
}
</style>
