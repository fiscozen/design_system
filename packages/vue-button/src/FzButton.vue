<template>
  <button type="button" :class="classes" @click="onClick" :style="style">{{ label }} </button>
</template>

<script lang="ts" setup>
import { computed } from 'vue';

const props = withDefaults(defineProps<{
  /**
   * The label of the button
   */
  label: string,
  /**
   * primary or secondary button
   */
  primary?: boolean,
  /**
   * size of the button
   */
  size?: 'small' | 'medium' | 'large',
  /**
   * background color of the button
   */
  backgroundColor?: string,

}>(), { primary: false });

const emit = defineEmits<{
  /**
   * simple emit function, always emits 1
   */
  (e: 'click', id: number): void;
}>();

const classes = computed(() => ({
  'storybook-button': true,
  'storybook-button--primary': props.primary,
  'storybook-button--secondary': !props.primary,
  [`storybook-button--${props.size || 'medium'}`]: true,
}));

const style = computed(() => ({
  backgroundColor: props.backgroundColor
}));

const onClick = () => {
  emit("click", 1)
};

</script>

<style>
.storybook-button {
  font-family: 'Nunito Sans', 'Helvetica Neue', Helvetica, Arial, sans-serif;
  font-weight: 700;
  border: 0;
  border-radius: 3em;
  cursor: pointer;
  display: inline-block;
  line-height: 1;
}
.storybook-button--primary {
  color: white;
  background-color: #1ea7fd;
}
.storybook-button--secondary {
  color: #333;
  background-color: transparent;
  box-shadow: rgba(0, 0, 0, 0.15) 0px 0px 0px 1px inset;
}
.storybook-button--small {
  font-size: 12px;
  padding: 10px 16px;
}
.storybook-button--medium {
  font-size: 14px;
  padding: 11px 20px;
}
.storybook-button--large {
  font-size: 16px;
  padding: 12px 24px;
}

</style>