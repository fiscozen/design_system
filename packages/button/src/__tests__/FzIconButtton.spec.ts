import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import FzButton from '../FzIconButton.vue'

describe('FzButton', () => {
  it('should match snapshot', () => {
    const wrapper = mount(FzButton, {
      props: {
        tooltip: 'some text',
        disabled: false,
        iconName: 'bell'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should emit native event', () => {
    const wrapper = mount(FzButton, {
      props: {
        tooltip: 'some text',
        disabled: false,
        iconName: 'bell'
      }
    })
    wrapper.trigger('click')

    const clickEvs = wrapper.emitted('click')
    expect(clickEvs).toHaveLength(1)
  })
  it('should not emit click when disabled', () => {
    const wrapper = mount(FzButton, {
      props: {
        tooltip: 'some text',
        disabled: true,
        iconName: 'bell'
      }
    })
    wrapper.trigger('click')

    const clickEvs = wrapper.emitted('click')
    // in questo caso emitted ritorna undefined e non un array vuoto
    expect(clickEvs).toBeUndefined()
  })
})
