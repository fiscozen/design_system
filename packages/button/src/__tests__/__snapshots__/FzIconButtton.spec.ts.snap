// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FzButton > should match snapshot 1`] = `
"<button type="button" class="relative rounded flex flex-shrink items-center justify-center font-medium border-1 border-transparent h-32 focus:border-blue-600 focus:border-solid focus:border-1 px-16 bg-blue-500 hover:bg-blue-600 disabled:bg-blue-200 text-core-white focus:bg-blue-500">
  <!--v-if-->
  <!--v-if--><span class="hidden h-0 w-0">some text</span>
</button>"
`;
